### PR TITLE
Implement money management mechanics.

### PR DESCRIPTION
**Pull request by calebkrauter**
1. Add Bill Flip Mechanic.
2. Add Bills Sliding Mechanic
3. Improve Bill Sliding Mechanic.
4. Update Bill Sliding Mechanic to be simpler and used for bundling only.
5. Update Viewport to Scale properly.
6. Remove Selector from Bills and Add Selector as its own scene with a camera to follow it.
7. Fix moving out of bounds bug.
8. Polish selector and bundle mechanic.
9. Add Bill Counting Mechanic with visual feedback.
10. Add Bundling Mechanic. 
11. Update Bundling Mechanic to not only miscount bills but also to gaslight player into thinking the correct amount of cash was counted.
12. Implement Keep Cash Mechanic.
13. Partial-Implementation Select+Bundle combination Mechanic to allow the player to both count and start/add to a bundle.
14. Update buttons to function based on "Press" rather than "Release".
- TODO: update Select+Bundle so that when you press bundle after starting a Select+Bundle cash bundle, the remaining cash to fill the bundle is added and a new bundle is started if there is overflow. If you did not count the cash in the bundle, you will be gaslit into thinking that the cash is correctly counted. The count will appear accurate and will likely be miscounted, a new bundle as a result may be started for overflow even if there is no overflow. Conversely, if all bills are counted in the bundle, the correct amount will be counted and added and if there is overflow that will also be accurate. 

If Bundle is pressed prior to a Select+Bundle combination, the Select+Bundle bill will be added to appear as accurately counted whether accounting for overflow or not. If the bundle is counted, it will also appear accurate and be accurate.

- TODO: add an indication for the user when Select is pressed to show that Bundle could also be pressed for a special action to count and start/add to a bundle.

- TODO: Refactor. I observed that I have been using semicolons occasionally without realizing it, I need to go through and remove or add them so that my code is consistent.
- Consider Refactor. It seems to be standard to use underscores for naming in GDScripts, I have been generally using lowerCamelCase for variables etc and underscores for functions. I would like to adhere to what is more common to GDScripts as it is easy to break consistency when Godot defaults variables to using underscores.

### CODE REVIEW

File: Zeffo/Scenes/Entities/Bundle2D.gd [hunk 1]
```diff
@@ -0,0 +1,11 @@
+extends AnimatedSprite2D
+
+
+# Called when the node enters the scene tree for the first time.
+func _ready():
```
Comment (coltenkrauter) [inline Zeffo/Scenes/Entities/Bundle2D.gd:5]: Why the underscore?

File: Zeffo/Scenes/Entities/bill.gd [hunk 2]
```diff
@@ -16,9 +18,12 @@ func get_denomination():
 func set_flipped(flipped_in):
 	self.isFlipped = flipped_in
 
+func is_flipped():
+	return self.isFlipped
 # Called when the node enters the scene tree for the first time.
 func _ready():
-	bill_Select.hide()
+	#bill_Select.hide()
```
Comment (coltenkrauter) [inline Zeffo/Scenes/Entities/bill.gd]:
```suggestion
```

File: Zeffo/Scenes/Entities/bundle.tscn [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+[sub_resource type="SpriteFrames" id="SpriteFrames_kyviv"]
+animations = [{
+"frames": [{
+"duration": 1.0,
+"texture": ExtResource("1_hpln2")
+}],
+"loop": true,
+"name": &"default",
+"speed": 5.0
+}]
```
Comment (coltenkrauter) [inline Zeffo/Scenes/Entities/bundle.tscn:16]: Is there a reason for not indenting?
```suggestion
animations = [{
    "frames": [{
        "duration": 1.0,
        "texture": ExtResource("1_hpln2")
    }],
    "loop": true,
    "name": &"default",
    "speed": 5.0
}]
```

File: Zeffo/Scenes/Entities/bundle.tscn [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+[node name="Bundle" type="Node2D"]
+script = ExtResource("1_rdrxn")
+
+[node name="Bundle2D" type="AnimatedSprite2D" parent="."]
+position = Vector2(-44, -322)
+scale = Vector2(0.2225, 0.2225)
+sprite_frames = SubResource("SpriteFrames_kyviv")
+script = ExtResource("3_jq5uy")
```
Comment (coltenkrauter) [inline Zeffo/Scenes/Entities/bundle.tscn:25]: Is this stuff auto-generated? Or are you following documentation or examples?

How did you know to do this?

File: Zeffo/Scenes/Entities/selEED4.tmp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+"loop": true,
+"name": &"default",
+"speed": 5.0
+}]
+
+[node name="Selector" type="Node2D"]
+
+[node name="CameraContainer" type="AnimatedSprite2D" parent="."]
+sprite_frames = SubResource("SpriteFrames_skcqn")
+animation = &""
```
Comment (coltenkrauter) [inline Zeffo/Scenes/Entities/selEED4.tmp:29]: Is this auto-generated?

**Review by coltenkrauter — COMMENTED** [review]
Can you share a bit more context about this code? 

Looks to me like there is auto-generated code mixed in with code you wrote. This is not ideal for collaboration if it is true.

**Comment by calebkrauter** [comment]
> Can you share a bit more context about this code? 
> 
> 
> 
> Looks to me like there is auto-generated code mixed in with code you wrote. This is not ideal for collaboration if it is true.

All TSCN files are autogenerated as they are scene files created by the game engine when you click and add nodes etc visually. All gdscript which are .gd files are written. I believe there is also a settings file that is autogenerated with .Godot appended. I'm not sure if there is a way to exclude the autogenerated files from  the PR review process while still including them in the PR which would streamline it.

File: Zeffo/Scenes/Entities/Bundle2D.gd (hunk 1 above)
Comment (calebkrauter) [reply Zeffo/Scenes/Entities/Bundle2D.gd:5]: This is a standard for functions that are default functions for the game engine's usage.

File: Zeffo/Scenes/Entities/bundle.tscn (hunk 3 above)
Comment (calebkrauter) [reply Zeffo/Scenes/Entities/bundle.tscn:16]: these are autogenerated files. I made a comment clarifying, and I may have mentioned this to you in a text previously. My bad for not clarifying in the description of the PR.

File: Zeffo/Scenes/Entities/bundle.tscn (hunk 4 above)
Comment (calebkrauter) [reply Zeffo/Scenes/Entities/bundle.tscn:25]: Yes.

File: Zeffo/Scenes/Entities/selEED4.tmp (hunk 5 above)
Comment (calebkrauter) [reply Zeffo/Scenes/Entities/selEED4.tmp:29]: Yes.

**Comment by coltenkrauter** [comment]
> > Can you share a bit more context about this code? 
> > 
> > 
> > 
> > Looks to me like there is auto-generated code mixed in with code you wrote. This is not ideal for collaboration if it is true.
> 
> All TCSN files are autogenerated as they are scene files created by the game engine when you click and add nodes etc visually. All gdscript which are .gd files are written. I believe there is also a settings file that is autogenerated with .Godot appended. I'm not sure if there is a way to exclude the autogenerated files from  the PR review process while still including them in the PR which would streamline it. 

Well, there are a few ways that you should be able to clean things up. One way is you could possibly use the  .gitignore file to ignore the auto generated file extensions. 

Another option, which might be preferred if it is possible, would be too configure a destination directory for auto generated files. For example, you could have a disc or output directory all the generated files going into. Possibly, you could gitignore the output directory.

But, I think that it's not elegant to have all of these auto generated files littered around the files that you have made code changes to.

**Comment by calebkrauter** [comment]
> > > Can you share a bit more context about this code?
> > > Looks to me like there is auto-generated code mixed in with code you wrote. This is not ideal for collaboration if it is true.
> > 
> > 
> > All TCSN files are autogenerated as they are scene files created by the game engine when you click and add nodes etc visually. All gdscript which are .gd files are written. I believe there is also a settings file that is autogenerated with .Godot appended. I'm not sure if there is a way to exclude the autogenerated files from  the PR review process while still including them in the PR which would streamline it.
> 
> Well, there are a few ways that you should be able to clean things up. One way is you could possibly use the .gitignore file to ignore the auto generated file extensions.
> 
> Another option, which might be preferred if it is possible, would be too configure a destination directory for auto generated files. For example, you could have a disc or output directory all the generated files going into. Possibly, you could gitignore the output directory.
> 
> But, I think that it's not elegant to have all of these auto generated files littered around the files that you have made code changes to.

Ok thanks for the response, my understanding with using .gitignore is that the files ignored will not be tracked by Git and will thus not be committed and pushed to Github. If this is correct, then adding these files to the .gitignore would prove dangerous. All or at least much of the auto-generated code is necessary for the game to function. All of the TSCN files which are auto-generated are modified anytime I make changes to the Scenes in the game engine. Anytime I make an entity or a grouping of Entities I make a Scene (TSCN) file where I can visually edit the Scene and the TCSN file is changed under the hood. When I make a new Scene it creates a new TSCN file. I need the TSCN files tracked by Git so that I have redundancy in the cloud. I find it unlikely that using .gitignore as convenient as it is, will be a good idea because the auto-generated files will only live on my local machine and could easily get lost or corrupted if something goes wrong meaning I won't be able to revert potentially vital changes. Similarly for the .Godot file which is used for settings, if I update project settings I am near certain that file is also updated under the hood and necessary to be tracked by Git. That being said, I may be able to include in the .gitignore the tmp and import files, however I'm not certain what affect that will have on my project or workflow, my main concern would be if I lost data on my project or it got corrupted, I may have to manually re-import images and I'm not sure what the significance of the tmp files are or when they are created.

I am happy to be proven wrong on this, maybe I have a misunderstanding of .gitignore or something else.

You second idea sounds promissing: "Another option, which might be preferred if it is possible, would be too configure a destination directory for auto generated files. For example, you could have a disc or output directory all the generated files going into. Possibly, you could gitignore the output directory."

Can you elaborate on this? I am not familiar. Are you referring to a disc/directory on my local machine, or is this perhaps a Github feature? It is unclear to me because you mention adding the directory/disc to the .gitignore which to me infers it would be local if I understand and thus I would encounter the problem I mentioned above where I have no data redundancy on the cloud for my Scene and Settings data.

Whatever the solution is, it likely needs to include at least the TSCN and .Godot auto generated files like you have encountered on the PR in the Git commits. In fact I'm already using a .gitignore template that is built for the game engine I am using, Godot4.

Also I totally agree that I'd like to clean up the PRs and make it easier to review, just not sure what the best way to do that is at the moment.

Currently I think the easiest option moving forward is for you to filter the files you review to only .gd files as those are the scripts I am writing. Then if you want you can just press view on the rest of them.

Let me know what your thoughts are.
Thanks!

**Comment by coltenkrauter** [comment]
> ...the auto-generated files will only live on my local machine and could easily get lost or corrupted...

Can't you just generate the files when they are needed? Why do they need to be stored in GitHub?

>  it likely needs to include at least the TSCN and .Godot auto generated files like you have encountered on the PR

Why? I often write code, such as typescript, and before getting deployed to production for customer use, it must get compiled (or auto-generate javascript files). I don't store these javascript files in GitHub... I generate them when they are needed.

**Comment by calebkrauter** [comment]
> > ...the auto-generated files will only live on my local machine and could easily get lost or corrupted...
> 
> 
> 
> Can't you just generate the files when they are needed? Why do they need to be stored in GitHub?
> 
> 
> 
> >  it likely needs to include at least the TSCN and .Godot auto generated files like you have encountered on the PR
> 
> 
> 
> Why? I often write code, such as typescript, and before getting deployed to production for customer use, it must get compiled (or auto-generate javascript files). I don't store these javascript files in GitHub... I generate them when they are needed.

I can look into that, but I think it's pretty unlikely. I don't think it's the same as adding in a library or dependencies. The TSCN files are part of the game creation. If I wasn't using a game engine, it would be actual code that I was writing in place of these files, but because the game engine editor streamlines the making of the game, I don't have to write all of the code, I just create some files which have code automatically generated from interpreting the changes I make visually in the editor. Then I attach .gd script files which I write and it tells the engine how to handle data and entities in the TSCN files as I understand it. There's probably a bit more to it than that. The files are auto generated in the sense that I don't have to write the code, but I suppose you could look at them like visual code files, where I visually change the underlying code.

If I get rid of the TSCN files, From my game, then the scenes will not appear, I won't have any entities, a player or UI or anything like that. So even though I'm not writing the code for those files, likely, they need to be treated as files that I wrote in regards to adding them to Git.

**Comment by calebkrauter** [comment]
> > ...the auto-generated files will only live on my local machine and could easily get lost or corrupted...
> 
> Can't you just generate the files when they are needed? Why do they need to be stored in GitHub?
> 
> > it likely needs to include at least the TSCN and .Godot auto generated files like you have encountered on the PR
> 
> Why? I often write code, such as typescript, and before getting deployed to production for customer use, it must get compiled (or auto-generate javascript files). I don't store these javascript files in GitHub... I generate them when they are needed.

I'll also mention, I took a look at the [GodotEngine Account](https://github.com/godotengine) at some PRs and it appears to me that it may be common to include many different file types including TSCN files.

**Comment by coltenkrauter** [comment]
If you get rid of the TSCN files, can they be replaced somehow?

**Comment by coltenkrauter** [comment]
Check this:

https://forum.godotengine.org/t/what-doesnt-need-to-be-committed-to-version-control-in-a-godot-project/28750

https://www.reddit.com/r/godot/comments/176z40o/exactly_which_folders_files_should_i_include_in/

It does look like a lot of folks include the auto-generated files..... I don't like this pattern much. 🤷🏻‍♂️

**Comment by calebkrauter** [comment]
> Check this:
> 
> 
> 
> https://forum.godotengine.org/t/what-doesnt-need-to-be-committed-to-version-control-in-a-godot-project/28750
> 
> 
> 
> https://www.reddit.com/r/godot/comments/176z40o/exactly_which_folders_files_should_i_include_in/
> 
> 
> 
> It does look like a lot of folks include the auto-generated files..... I don't like this pattern much. 🤷🏻‍♂️ 

Gotcha. Looks like ignoring .import and .tmp could be okay.

**Comment by calebkrauter** [comment]
> If you get rid of the TSCN files, can they be replaced somehow?

Unfortunately I'm pretty certain that it's not possible to regenerate the TSCN files. I think it is highly unlikely. It would be like getting rid of half of the game project. The TSCN files are something you create using the editor though the code is automatically generated from interpreting your changes. It's not like you can just default each of them because that would just be an empty TSCN file. Scene files are like building blocks similar but different to Objects because you can make many Scenes with lots of state, images, or other attributes and then nest Scenes in other Scenes to build levels, characters, weapons or anything you can imagine. So they are automatically generated in the sense that you didn't write the code. Not in the sense that you can generate all that you need from a Scene if you delete a Scene. 

It might be easier to show you how it works over Discord if you'd like more clarity on the situation. I feel that I've explained the situation pretty thoroughly though I can try to explain more if you like. Let me know if you'd like to connect and thanks for the feedback.

**Review by coltenkrauter — APPROVED** [review]

**Pull request merged**